qiskit_nature/utils/linalg.py

```diff
@@ -0,0 +1,176 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Linear algebra utilities."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, List, Sequence, Tuple, Union
+
+import numpy as np
+from qiskit import QuantumRegister
+from qiskit.circuit import Gate, Qubit
+from qiskit.circuit.library import RZGate, XGate, XXPlusYYGate
+
+# HACK: Sphinx fails to handle "ellipsis"
+# See https://github.com/python/typing/issues/684
+if TYPE_CHECKING:
+    _SliceAtom = Union[int, slice, "ellipsis"]
+else:
+    _SliceAtom = Union[int, slice, type(Ellipsis)]
+
+_Slice = Union[_SliceAtom, Tuple[_SliceAtom, ...]]
+
+
+def apply_matrix_to_slices(
+    target: np.ndarray, matrix: np.ndarray, slices: Sequence[_Slice]
+) -> np.ndarray:
+    """Apply a matrix to slices of a target tensor.
+
+    Args:
+        target: The tensor containing the slices on which to apply the matrix.
+        matrix: The matrix to apply to slices of the target tensor.
+        slices: The slices of the target tensor on which to apply the matrix.
+
+    Returns:
+        The tensor resulting from applying the matrix to the slices of the target tensor.
+    """
+    result = target.copy()
+    for i, slice_i in enumerate(slices):
+        result[slice_i] *= matrix[i, i]
+        for j, slice_j in enumerate(slices):
+            if j != i:
+                result[slice_i] += target[slice_j] * matrix[i, j]
+    return result
+
+
+def givens_matrix(a: Union[complex, float], b: Union[complex, float]) -> np.ndarray:
```

Comment:
I don't know what you mean by "encompasses". I think floats should be allowed here.

Reply:
Oh interesting, mypy seems fine with this. I didn't know that would work, it seems like it should be a TypeError since floats are not instances of complex.